src/python/turicreate/data_structures/sarray.py

```diff
@@ -355,6 +356,11 @@ def __init__(self, data=[], dtype=None, ignore_cast_failure=False, _proxy=None):
                 self.__proxy__ = data.astype(dtype).__proxy__
         else:
             self.__proxy__ = UnitySArrayProxy()
+
+            ## data transfromation from generator to list
+            if (isinstance(data, types.GeneratorType)) or (sys.version_info.major >= 3 and isinstance(data, (filter, map))):
```

Comment:
Should we add `range` to the `(filter, map)` tuple?

Reply:
surprisingly the range is not a generator. if we want to support range, then we should also support xrange in py2? Besides, [sarrary.from_sequence](https://apple.github.io/turicreate/docs/api/generated/turicreate.SArray.from_sequence.html#turicreate.SArray.from_sequence) have a more efficient representation of the data structure as it doesn't need to materialize the range into a list until evaluation time.

Reply:
Supporting `xrange` in py2.7 is a great idea. 

I agree it would be better for our users to use `sarray.from_sequence(...)` but I think we should also support range here. Also `sarray.from_sequence(...)` doesn't have a `step` parameter like the built in `range` so it may not work for their use case.

Reply:
yeah! Good point! I will add this on!